src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -65,9 +65,10 @@ protected function buildRoleMap()
                 }
 
                 $visited[] = $role;
-                $this->map[$main] = array_unique(array_merge($this->map[$main], $this->hierarchy[$role]));
+                $this->map[$main] = array_merge($this->map[$main], $this->hierarchy[$role]);
                 $additionalRoles = array_merge($additionalRoles, array_diff($this->hierarchy[$role], $visited));
             }
+            $this->map[$main] = array_unique($this->map[$main]);
         }
```

Comment:
Those nested `array_merges` could really slow things down as well. In 5.6 I would do this:

```php
foreach (...) {
    $maps[] = $this->hierarchy[$role];
}
$this->map[$main] = array_unique(array_merge($this->map[$main], ...$maps));
```
I'm sure there's a 5.5 equivalent, it should speed up things even more due to less array merges (same could be for `$additionalRoles`).